relayer/src/link/relay_path.rs

```diff
@@ -1387,22 +1425,36 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> RelayPath<ChainA, ChainB> {
         }
 
         // Update clients ahead of scheduling the operational data, if the delays are non-zero.
-        if !self.zero_delay() {
-            debug!("connection delay is non-zero: updating client");
+        od.scheduled_time = if self.conn_delay_needed(&od) {
+            debug!("connection delay must be taken into account: updating client");
             let target_height = od.proofs_height.increment();
-            match od.target {
+            let update_time = match od.target {
                 OperationalDataTarget::Source => {
-                    self.update_client_src(target_height, &od.tracking_id)?
+                    let update_height = self.update_client_src(target_height, &od.tracking_id)?;
+                    self.src_chain()
+                        .query_host_consensus_state(update_height)
+                        .map_err(LinkError::relayer)?
+                        .timestamp()
                 }
                 OperationalDataTarget::Destination => {
-                    self.update_client_dst(target_height, &od.tracking_id)?
+                    let update_height = self.update_client_dst(target_height, &od.tracking_id)?;
+                    self.dst_chain()
+                        .query_host_consensus_state(update_height)
+                        .map_err(LinkError::relayer)?
+                        .timestamp()
                 }
             };
+            let duration = Timestamp::now()
+                .duration_since(&update_time)
+                .unwrap_or(Duration::ZERO);
+            Instant::now().sub(duration)
```

Comment:
IMHO, `Instant::now()` is a better choice v/s the chain's time (i.e. latest block's timestamp), this is because the chain handler will ultimately check this logic against the timestamp of the next block (which will contain this tx/message). 
One problem with this choice is that it might lead to false positives (with the right timing and drift). (I believe https://github.com/informalsystems/ibc-rs/issues/1868 is one such case)

Reply:
I think this change is ok but not clear to me it is needed to solve the actual issue. Aren't we supposed to change this line further down in `try_fetch_scheduled_operational_data`:
`op.scheduled_time.elapsed() > connection_delay`

and get the time since the last host consensus state instead of `elapsed()`?

Reply:
The idea was to set the `scheduled_time` to a time in the past when the header was updated (i.e. `Instant::now().sub(duration)` where duration is the timestamp of the block that includes the header update). So, `op.scheduled_time.elapsed()` returns the time elapsed since the header update. Also, I thought the chain handler will check this against the block's timestamp which contains this packet, so `now()` is a better choice than the latest host timestamp. Not sure if my reasoning is correct. 🤔

Reply:
I think this part is correct. But #1757 is about this:
```
The relayer logic that implements connection delay support checks packets against the relayer's local time. This is inconsistent with the IBC handlers implementation that checks against the block time of the latest on-chain block.
```

It refers to the packet verification potentially failing. If the update client was included in block `h` and the last block is `H`, the packet handler for a Tx to be included in `H+1` will check that `H.time - h.time` is bigger than connection delay. But in `try_fetch_scheduled_operational_data()` we basically check against `Instant::now() - h.time`. So we may send the packet too early.

We have fixed a similar issue when deciding if a packet has timed out https://github.com/informalsystems/ibc-rs/blob/f11415afbf583124d1a3ae52c0a972a7680bd716/modules/src/core/ics04_channel/packet.rs#L167

Note: To complicate things, there is an additional check in the packet handler that asserts that a minimum number `n` of blocks must have been created since the update client block. `n` is computed as `ceil(connection_delay/max_expected_time_per_block)`. We should then do some basic check along these lines in the relayer code. `max_expected_time_per_block` can be found in the genesis info:
```
{
  ...
  "app_state": {
    ...
    "ibc": {
      ...
      "connection_genesis": {
        ...
        "params": {
          "max_expected_time_per_block": "30000000000"
        }
...
```

Reply:
> Also, thanks for reminding me about the block delay 👌, I had opened issue #1773 for that. Should we address that as well (as part of this PR)?

Ideally here unless is a complex change.

Reply:
> I believe #1868 is one such case

Looked through that but it looks like 0 connection delay there. So probably it's a different issue.

Reply:
I think we were seeing a similar issue in the integration test PR #1950. The fix was to increase the `connection_delay` by `avg_block_time` of the host so that we are sure that the packet verification succeeds -> https://github.com/informalsystems/ibc-rs/pull/1950/commits/19fd37f05c07faf216c6b25b6ed1680c212c4b97
Is there a way to get a chain's avg block time (using RPC)?

Reply:
> > My understanding was that the handler does this instead -> `((H + 1).time - h.time) > conn_delay`
> 
> Ah yes, you are correct! forgot context is set in `BeginBlock` (so before `deliverTx()`)

So actually this is true for `deliverTx()` handling. However when we do the Tx simulation, the checks run on the simulate state copy at latest height `H` which may have been created on chain before the time we checked for elapse of the delay. This is why test fails. 

Your changes in 19fd37f05c07faf216c6b25b6ed1680c212c4b97 are then fine, we just need to get `max_expected_time_per_block` from the genesis parameters instead of the constant `AVG_BLOCK_TIME_SECS`. 
Given that this is a chain parameter I am wondering if we can get/cache it when we create the chain handler.

We are probably just lucky that the test passes in master as we mainly test with small block times (1sec) and because we use `Instant::now()` for `scheduled_time` which is more recent than the block containing the client update...so we end up delaying more.

Reply:
Thinking about this more, maybe it's still better to go with https://github.com/informalsystems/ibc-rs/pull/1917#discussion_r825734002. So we get the `last_header_time` (using the same `query_host_consensus_state()`) and then something like:
`last_header_time.sub(op.scheduled_time) > connection_delay`

with the weird way of moving between `Timestamp` and `Instant` :( Maybe have some wrappers for these 

Dealing with block times only will avoid additional failure scenarios caused by clock accuracy differences between hermes and the "chain" (header BFT time or proposer time in the future).

Reply:
> However when we do the Tx simulation, the checks run on the simulate state copy at latest height H which may have been created on chain before the time we checked for elapse of the delay. This is why test fails.

Wow! That's a tricky one, nice catch! 👍 

> Your changes in https://github.com/informalsystems/ibc-rs/commit/19fd37f05c07faf216c6b25b6ed1680c212c4b97 are then fine, we just need to get max_expected_time_per_block from the genesis parameters instead of the constant AVG_BLOCK_TIME_SECS.

I checked the ibc-go codebase and they seem to view the `max_expected_time_per_block` differently. The default value is 30 seconds and they view it as the max time a chain might take to mint a block. (https://github.com/cosmos/ibc-go/pull/171#discussion_r630932520) Should we still use it? Or is there an average block time counterpart?

Reply:
There is no average block time and even if it were it wouldn't always work. Using max also delays relaying quite a bit. To avoid this we can just use the latest block time as I mentioned in the last comment.

Reply:
Great idea! I tried to do this with https://github.com/informalsystems/ibc-rs/pull/1917/commits/4df249a22f9754a343293df805d3804853da814d.

Reply:
Maybe we can cache this at some point.